In [1]:
import graphlab

A newer version of GraphLab Create (v1.8.4) is available! Your current version is v1.8.2.

You can use pip to upgrade the graphlab-create package. For more information see https://dato.com/products/create/upgrade.


In [2]:
people = graphlab.SFrame('people_wiki.gl/')

[INFO] GraphLab Create v1.8.2 started. Logging: C:\Users\Thien\AppData\Local\Temp\graphlab_server_1457792840.log.0


In [ ]:
graphlab.SFrame()

In [7]:
john = people[people['name'] == 'Elton John']

In [8]:
john

Columns:
	URI	str
	name	str
	text	str

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use len(sf) to force materialization.

In [9]:
john['word_count'] = graphlab.text_analytics.count_words(john['text'])

In [10]:
john['word_count']

dtype: dict
Rows: 1
[{'all': 1L, 'six': 1L, 'producer': 1L, 'heavily': 1L, 'over': 2L, 'named': 1L, 'fifty': 1L, 'four': 1L, 'openly': 1L, 'including': 1L, 'highestprofile': 1L, 'years': 1L, 'its': 2L, 'impact': 1L, 'westminster': 1L, '27': 1L, '21': 2L, 'wed': 1L, 'had': 1L, '1947': 1L, 'abbey': 1L, 'winning': 1L, 'late': 1L, 'to': 4L, 'commander': 1L, 'about': 1L, 'born': 1L, '2014': 1L, 'as': 2L, 'has': 9L, '2013': 1L, 'his': 4L, 'march': 1L, 'than': 3L, 'song': 1L, 'songwriter': 2L, 'continues': 1L, 'records': 1L, 'five': 1L, 'occasional': 1L, 'they': 1L, 'inception': 1L, 'world': 1L, 'brit': 1L, 'him': 3L, 'datein': 1L, 'hall': 2L, 'fivedecade': 1L, 'knighthood': 1L, 'bestselling': 2L, 'artist': 1L, 'be': 1L, '1996': 1L, 'list': 1L, 'roll': 2L, 'hercules': 1L, 'announced': 1L, 'rock': 2L, 'become': 1L, 'bernie': 1L, 'outstanding': 1L, 'england': 1L, 'composer': 1L, 'queens': 1L, 'foundation': 2L, 'diana': 1L, 'globe': 1L, 'artists': 2L, 'culture': 1L, 'been': 3L, '49': 1L, 'year':

In [11]:
john_wordcount_table = john[['word_count']].stack("word_count", new_column_name = ['word','count'])

In [12]:
john_wordcount_table

Columns:
	word	str
	count	int

Rows: 255

Data:
+-----------+-------+
|    word   | count |
+-----------+-------+
| movements |   1   |
|   social  |   1   |
|  champion |   1   |
|    wed    |   1   |
|   legal   |   1   |
|   became  |   1   |
|   after   |   1   |
|    2005   |   1   |
|  december |   2   |
|  furnish  |   2   |
+-----------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [13]:
john_wordcount_table.sort('count',ascending=False)

Columns:
	word	str
	count	int

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   27  |
|   in  |   18  |
|  and  |   15  |
|   of  |   13  |
|   a   |   10  |
|  has  |   9   |
|   he  |   7   |
|  john |   7   |
|   on  |   6   |
| since |   5   |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### Top word count words for Elton John 
The in and

# Compute TF-IDF for the corpus 

To give more weight to informative words, we weigh them by their TF-IDF scores.

In [14]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])

In [15]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'since': 1L, 'carltons': ... |
| alfred j lewy aka sandy le... | {'precise': 1L, 'thomas': ... |
| harpdog brown is a singer ... | {'just': 1L, 'issued': 1L,... |
| franz rottensteiner born i... | {'all': 1L, 'bauforschung'... |
| henry krvits born 30 decem... | {'legendary': 1L, 'gangste... |
| sam henderson born october... | {'now': 1L, 'currently': 1... |
| aaron lacrate is an americ... | {'exclusive': 2L, 'produce... |
| trevor ferguson aka john f... | {'taxi': 1L, 'salon': 1L, ... |
| grant nelson born 27 april... | {'houston': 1L, 'frankie':... |
| cathy caruth born 1955 is ... | {'phenomenon': 1L, 'debora... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

In [16]:
people['tfidf'] = graphlab.text_analytics.tf_idf(people['word_count'])

In [33]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'since': 1L, 'carltons': ... |
| alfred j lewy aka sandy le... | {'precise': 1L, 'thomas': ... |
| harpdog brown is a singer ... | {'just': 1L, 'issued': 1L,... |
| franz rottensteiner born i... | {'all': 1L, 'bauforschung'... |
| henry krvits born 30 decem... | {'legendary': 1L, 'gangste... |
| sam henderson born october... | {'now': 1L, 'currently': 1... |
| aaron lacrate is an americ... | {'exclusive': 2L, 'produce... |
| trevor ferguson aka john f... | {'taxi': 1L, 'salon': 1L, ... |
| grant nelson born 27 april... | {'houston': 1L, 'frankie':... |
| cathy caruth born 1955 is ... | {'phenomenon': 1L, 'debora... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'since': 1.45537671730804... |
| {'precise': 6.443200606955... |
| {'just': 2.700729968710864... |
| {'all': 1.6431112434912472... |
| {'legendary': 4.2808562943... |
| {'now': 1.96695239252401, ... |
| {'exclusive': 10.455187230... |
| {'taxi': 6.052021456094502... |
| {'houston': 3.935505942157... |
| {'phenomenon': 5.750053426... |
+-------------------------------+
[10 rows x 5 columns]

In [18]:
john = people[people['name'] == 'Elton John']

In [22]:
john[['tfidf']].stack("tfidf",new_column_name = ['word','tfidf']).sort('tfidf',ascending = False)

Columns:
	word	str
	tfidf	float

Rows: 255

Data:
+---------------+---------------+
|      word     |     tfidf     |
+---------------+---------------+
|    furnish    |  18.38947184  |
|     elton     |  17.48232027  |
|   billboard   | 17.3036809575 |
|      john     | 13.9393127924 |
|  songwriters  |  11.250406447 |
|  overallelton | 10.9864953892 |
| tonightcandle | 10.9864953892 |
|    19702000   | 10.2933482087 |
|   fivedecade  | 10.2933482087 |
|      aids     |  10.262846934 |
+---------------+---------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### Top TF-IDF words for Elton John
Furnish elton billboard

### Manually compute distances between a few people

In [23]:
john = people[people['name'] == 'Elton John']

In [24]:
Victoria = people[people['name'] == 'Victoria Beckham']

In [25]:
graphlab.distances.cosine(john['tfidf'][0],Victoria['tfidf'][0])

0.9567006376655429

In [26]:
Paul = people[people['name'] == 'Paul McCartney']

In [27]:
graphlab.distances.cosine(john['tfidf'][0],Paul['tfidf'][0])

0.8250310029221779

### The cosine distance between 'Elton John's and 'Victoria Beckham's articles (represented with TF-IDF) falls within which range? 0.9567006376655429 

### The cosine distance between 'Elton John's and 'Paul McCartney's articles (represented with TF-IDF) falls within which range? 0.8250310029221779

### Who is closer to 'Elton John', 'Victoria Beckham' or 'Paul McCartney'? Victoria Beckham

# Nearest neighbor model for document retrieval

In [58]:
tfidfmodel = graphlab.nearest_neighbors.create(people, features=['tfidf'],label = 'name',distance='cosine')

Starting brute force nearest neighbors model training.

In [59]:
wordcountmodel = graphlab.nearest_neighbors.create(people,features=['word_count'], label = 'name', distance='cosine')

Starting brute force nearest neighbors model training.

In [60]:
john = people[people['name'] == 'Elton John']

In [62]:
wordcountmodel.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 7.51ms       |

| Done         |         | 100         | 330.746ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-------------------+------+
| query_label |  reference_label  |      distance     | rank |
+-------------+-------------------+-------------------+------+
|      0      |     Elton John    | 2.22044604925e-16 |  1   |
|      0      |   Cliff Richard   |   0.16142415259   |  2   |
|      0      |   Sandro Petrone  |   0.16822542751   |  3   |
|      0      |    Rod Stewart    |   0.168327165587  |  4   |
|      0      | Malachi O'Doherty |   0.177315545979  |  5   |
+-------------+-------------------+-------------------+------+
[5 rows x 4 columns]

### Who is the nearest neighbor to 'Elton John' using raw word counts? Cliff Richard

In [61]:
tfidfmodel.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 9.515ms      |

| Done         |         | 100         | 323.239ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+--------------------+------+
| query_label | reference_label  |      distance      | rank |
+-------------+------------------+--------------------+------+
|      0      |    Elton John    | -2.22044604925e-16 |  1   |
|      0      |   Rod Stewart    |   0.717219667893   |  2   |
|      0      |  George Michael  |   0.747600998969   |  3   |
|      0      | Sting (musician) |   0.747671954431   |  4   |
|      0      |   Phil Collins   |   0.75119324879    |  5   |
+-------------+------------------+--------------------+------+
[5 rows x 4 columns]

### Who is the nearest neighbor to 'Elton John' using TF-IDF? Rod Stewart

In [63]:
Victoria = people[people['name'] == 'Victoria Beckham'] 

In [64]:
wordcountmodel.query(Victoria)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 8.01ms       |

| Done         |         | 100         | 313.731ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+--------------------+------+
| query_label |     reference_label      |      distance      | rank |
+-------------+--------------------------+--------------------+------+
|      0      |     Victoria Beckham     | -2.22044604925e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |   0.207307036115   |  2   |
|      0      |      Adrienne Corri      |   0.214509782788   |  3   |
|      0      |     Beverly Jane Fry     |   0.217466468741   |  4   |
|      0      |      Raman Mundair       |   0.217695474992   |  5   |
+-------------+--------------------------+--------------------+------+
[5 rows x 4 columns]

### Who is the nearest neighbor to 'Victoria Beckham' using raw word counts? Mary Fitzgerald (artist)

In [65]:
tfidfmodel.query(Victoria)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 13.007ms     |

| Done         |         | 100         | 336.75ms     |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+---------------------+-------------------+------+
| query_label |   reference_label   |      distance     | rank |
+-------------+---------------------+-------------------+------+
|      0      |   Victoria Beckham  | 1.11022302463e-16 |  1   |
|      0      |    David Beckham    |   0.548169610263  |  2   |
|      0      | Stephen Dow Beckham |   0.784986706828  |  3   |
|      0      |        Mel B        |   0.809585523409  |  4   |
|      0      |    Caroline Rush    |   0.819826422919  |  5   |
+-------------+---------------------+-------------------+------+
[5 rows x 4 columns]

### Who is the nearest neighbor to 'Victoria Beckham' using TF-IDF? David Beckham